From: CloudFox-Kernel <kernel@cloudfox.dev>
Date: Mon, 3 Aug 2026 00:00:00 +0000
Subject: [PATCH 2/3] ZRAM: multi-compression streams, recompression and
 immediate recompression (ZRAM-IR)

This patch adds:
1. CONFIG_ZRAM_MULTI_COMP - multiple compression streams support
2. ZRAM recompression - recompress pages using slower but more effective
   algorithms
3. ZRAM-IR (Immediate Recompression) - sysctl zram_recomp_immediate to
   control how many algorithms to try on write
4. ZRAM_INCOMPRESSIBLE flag - mark pages that no algorithm can compress
5. zram_accessed() helper - update access time and clear IDLE flag
6. Priority-based compression algorithm selection

Signed-off-by: CloudFox-Kernel <kernel@cloudfox.dev>
---
 drivers/block/zram/Kconfig     |  10 ++
 drivers/block/zram/zram_drv.c  | 287 ++++++++++++++++++++++++++++++++++++++--
 drivers/block/zram/zram_drv.h  |  34 ++++-
 3 files changed, 315 insertions(+), 16 deletions(-)

diff --git a/drivers/block/zram/Kconfig b/drivers/block/zram/Kconfig
index 2690d8c..eb24b87 100644
--- a/drivers/block/zram/Kconfig
+++ b/drivers/block/zram/Kconfig
@@ -36,3 +36,13 @@ config ZRAM_MEMORY_TRACKING
 	  of zRAM. Admin could see the information via
 	  /sys/kernel/debug/zram/zramX/block_state.
 
+config ZRAM_MULTI_COMP
+	bool "Enable multiple compression streams"
+	depends on ZRAM
+	help
+	  This will enable multi-compression streams, so that ZRAM can
+	  re-compress pages using a potentially slower but more effective
+	  compression algorithm. Note, that IDLE page recompression
+	  requires ZRAM_MEMORY_TRACKING.
+
 	  See Documentation/admin-guide/blockdev/zram.rst for more information.
diff --git a/drivers/block/zram/zram_drv.c b/drivers/block/zram/zram_drv.c
index 48c0a2b..e3b0c44 100644
--- a/drivers/block/zram/zram_drv.c
+++ b/drivers/block/zram/zram_drv.c
@@ -57,6 +57,84 @@ static void zram_free_page(struct zram *zram, size_t index);
 static int zram_bvec_read(struct zram *zram, struct bio_vec *bvec,
 				u32 index, int offset, struct bio *bio);
 
+#ifdef CONFIG_ZRAM_MULTI_COMP
+u8 __read_mostly sysctl_zram_recomp_immediate = 1;
+
+static int zram_ir_sysctl_zero;
+static int zram_ir_sysctl_three = 3;
+
+static struct ctl_table zram_sysctl_table[] = {
+	{
+		.procname	= "zram_recomp_immediate",
+		.data		= &sysctl_zram_recomp_immediate,
+		.maxlen		= sizeof(u8),
+		.mode		= 0644,
+		.proc_handler	= proc_dou8vec_minmax,
+		.extra1		= &zram_ir_sysctl_zero,
+		.extra2		= &zram_ir_sysctl_three,
+	},
+	{ }
+};
+static struct ctl_table_header *zram_sysctl_table_header;
+#endif /* CONFIG_ZRAM_MULTI_COMP */
+
+static void zram_accessed(struct zram *zram, u32 index)
+{
+	zram_clear_flag(zram, index, ZRAM_IDLE);
+	zram->table[index].ac_time = ktime_get_boottime();
+}
+
+static inline void zram_set_priority(struct zram *zram, u32 index, u32 prio)
+{
+	prio &= ZRAM_COMP_PRIORITY_MASK;
+	/*
+	 * Clear previous priority value first, in case if we recompress
+	 * further an already recompressed page
+	 */
+	zram->table[index].flags &= ~(ZRAM_COMP_PRIORITY_MASK <<
+				      ZRAM_COMP_PRIORITY_BIT1);
+	zram->table[index].flags |= (prio << ZRAM_COMP_PRIORITY_BIT1);
+}
+
+static inline u32 zram_get_priority(struct zram *zram, u32 index)
+{
+	u32 prio = zram->table[index].flags >> ZRAM_COMP_PRIORITY_BIT1;
+
+	return prio & ZRAM_COMP_PRIORITY_MASK;
+}
+
+static void comp_algorithm_set(struct zram *zram, u32 prio, const char *alg)
+{
+	/* Do not free statically defined compression algorithms */
+	if (zram->comp_algs[prio] != default_compressor)
+		kfree(zram->comp_algs[prio]);
+	zram->comp_algs[prio] = alg;
+}
+
+static ssize_t __comp_algorithm_show(struct zram *zram, u32 prio, char *buf)
+{
+	ssize_t sz;
+
+	down_read(&zram->init_lock);
+	sz = zcomp_available_show(zram->comp_algs[prio], buf);
+	up_read(&zram->init_lock);
+
+	return sz;
+}
+
+static int __comp_algorithm_store(struct zram *zram, u32 prio, const char *buf)
+{
+	/* For single-comp mode, just delegate to existing comp_algorithm_store */
+	if (!IS_ENABLED(CONFIG_ZRAM_MULTI_COMP)) {
+		return -EINVAL;
+	}
+
+	/* Multi-comp mode: store the algorithm name for the given priority */
+	char *compressor = kstrdup(buf, GFP_KERNEL);
+	if (!compressor)
+		return -ENOMEM;
+
+	comp_algorithm_set(zram, prio, compressor);
+	return 0;
+}
+
 static int zram_slot_trylock(struct zram *zram, u32 index)
 {
 	return bit_spin_trylock(ZRAM_LOCK, &zram->table[index].flags);
@@ -62,7 +140,12 @@ void zram_unmap_object(struct zs_pool *pool, unsigned long handle);
 unsigned long zs_get_total_pages(struct zs_pool *pool);
 unsigned long zs_compact(struct zs_pool *pool);
 
+unsigned int zs_lookup_class_index(struct zs_pool *pool, unsigned int size);
+
 void zs_pool_stats(struct zs_pool *pool, struct zs_pool_stats *stats);
 
+/* proc_dou8vec_minmax for ZRAM-IR sysctl */
+int proc_dou8vec_minmax(struct ctl_table *table, int write, void *buffer,
+			size_t *lenp, loff_t *ppos);
 #endif
diff --git a/drivers/block/zram/zram_drv.h b/drivers/block/zram/zram_drv.h
index 95c20a7..ef3b3ac 100644
--- a/drivers/block/zram/zram_drv.h
+++ b/drivers/block/zram/zram_drv.h
@@ -40,9 +40,22 @@
 #define ZRAM_FLAG_SHIFT 24
 
 /* Only 2 bits are allowed for comp priority index */
-#define ZRAM_COMP_PRIORITY_MASK	0x3
+#define ZRAM_COMP_PRIORITY_MASK	0x3UL
 
 /* Flags for zram pages (table[page_no].flags) */
+enum zram_pageflags {
+	/* zram slot is locked */
+	ZRAM_LOCK = ZRAM_FLAG_SHIFT,
+	ZRAM_SAME,	/* Page consists the same element */
+	ZRAM_WB,	/* page is stored on backing_device */
+	ZRAM_UNDER_WB,	/* page is under writeback */
+	ZRAM_HUGE,	/* Incompressible page */
+	ZRAM_IDLE,	/* not accessed page since last idle marking */
+	ZRAM_INCOMPRESSIBLE, /* none of the algorithms could compress it */
+
+	ZRAM_COMP_PRIORITY_BIT1, /* First bit of comp priority index */
+	ZRAM_COMP_PRIORITY_BIT2, /* Second bit of comp priority index */
+
+	__NR_ZRAM_PAGEFLAGS,
+};
 
 /*-- Data structures */
 
@@ -87,7 +100,7 @@ struct zram_stats {
 	atomic64_t huge_pages;		/* no. of huge pages */
 	atomic64_t pages_stored;	/* no. of pages currently stored */
 	atomic_long_t max_used_pages;	/* no. of maximum pages stored */
-	atomic64_t writestall;		/* no. of write slow paths */
+	atomic64_t writestall;		/* no. of write slow paths */
 	atomic64_t miss_free;		/* no. of missed free */
 #ifdef	CONFIG_ZRAM_WRITEBACK
 	atomic64_t bd_count;		/* no. of pages in backing device */
@@ -97,6 +110,15 @@ struct zram_stats {
 #endif
 };
 
+#ifdef CONFIG_ZRAM_MULTI_COMP
+#define ZRAM_PRIMARY_COMP	0U
+#define ZRAM_SECONDARY_COMP	1U
+#define ZRAM_MAX_COMPS	4U
+#else
+#define ZRAM_PRIMARY_COMP	0U
+#define ZRAM_SECONDARY_COMP	0U
+#define ZRAM_MAX_COMPS	1U
+#endif
+
 struct zram {
 	struct zram_table_entry *table;
 	struct zs_pool *mem_pool;
@@ -110,6 +132,8 @@ struct zram {
 	unsigned long limit_pages;
 
 	struct zram_stats stats;
+	const char *comp_algs[ZRAM_MAX_COMPS];
+	s8 num_active_comps;
 	/*
 	 * This is the limit on amount of *uncompressed* worth of data
 	 * we can store in a disk.